app/domain/authentication/authn_iam/authenticator.rb

```diff
@@ -0,0 +1,84 @@
+require 'active_support'
+require 'active_support/core_ext'
+require 'net/ldap'
+require 'json'
+require 'aws-sdk-iam'
+require 'aws-sdk-core'
+
+
+module Authentication
+  module AuthnIam
+
+    class Authenticator
+      
+      
+      def initialize(env:)
+        @env = env
+      end
+
+      def valid?(input)
+        @authn_name, @account, @login, password, @service_id = input.authenticator_name, input.account, input.username, input.password, input.service_id
+
+        # JSON holding the AWS signed headers 
+        @signed_aws_headers = JSON.parse input.password
+
+        is_trusted_by_aws? && (iam_role_matches? host_role)
+
+      end
+
+      def is_trusted_by_aws?
+
+        url = 'https://sts.amazonaws.com/?Action=GetCallerIdentity&Version=2011-06-15'
+
+        # Executes the AWS Signed Request
+        uri = URI.parse(url)
+        https = Net::HTTP.new(uri.host,uri.port)
+        https.use_ssl = true
+        https.set_debug_output($stdout)
+        request = Net::HTTP::Get.new(url)
+
+        @signed_aws_headers.each do |key, value|
+          request.add_field(key, value)
+        end
+
+        res = https.request(request)
+
+        Rails.logger.info("request => #{request}")
+        Rails.logger.info("****> #{res.code} #{res.message}")
+        Rails.logger.info("**** Body -> #{res.body} ")
+
+        @aws_response_hash = Hash.from_xml(res.body) if res.code.eql? "200"
+
+        res.code.eql? "200"
+
+      end
+    
+      def iam_role_matches? resource
```

Comment:
If this method takes both a resource and response hash, it'll be easy to write a unit test to verify it works as expected.

Reply:
Incorporated, and refactored